compiler_opt/distributed/local/local_worker_manager.py

```diff
@@ -0,0 +1,230 @@
+    if task.is_urgent:
+      try:
+        res = application()
+        out_q.put(TaskResult(msgid=task.msgid, success=True, value=res))
+      except BaseException as e:  # pylint: disable=broad-except
+        out_q.put(TaskResult(msgid=task.msgid, success=False, value=e))
+    else:
+      pool.submit(application).add_done_callback(make_ondone(task.msgid))
+
+
+def _run(*args, **kwargs):
+  try:
+    _run_impl(*args, **kwargs)
+  except BaseException as e:
+    logging.error(e)
+    raise e
+
+
+def _make_stub(cls: 'type[Worker]', *args, **kwargs):
+
+  class _Stub():
+    """Client stub to a worker hosted by a process."""
+
+    def __init__(self):
+      self._send: 'queue.Queue[Task]' = multiprocessing.get_context().Queue()
+      self._receive: 'queue.Queue[Optional[TaskResult]]' = \
+        multiprocessing.get_context().Queue()
+
+      # this is the process hosting one worker instance.
+      self._process = multiprocessing.Process(
+          target=functools.partial(
+              _run,
+              worker_class=cls,
+              in_q=self._send,
+              out_q=self._receive,
+              *args,
+              **kwargs))
+      # lock for the msgid -> reply future map. The map will be set to None
+      # when we stop.
+      self._lock = threading.Lock()
+      self._map: Dict[int, concurrent.futures.Future] = {}
+
+      # thread drainig the receive queue
+      self._pump = threading.Thread(target=self._msg_pump)
+      def observer():
+        self._process.join()
+        self._receive.put(None)
+      self._observer = threading.Thread(target=observer)
+
+      # atomic control to _msgid
+      self._msgidlock = threading.Lock()
+      self._msgid = 0
+
+      # start the worker and the message pump
+      self._process.start()
+      # the observer must follow the process start, otherwise join() raises.
+      self._observer.start()
+      self._pump.start()
+
+    def _msg_pump(self):
+      while True:
+        task_result = self._receive.get()
+        if not task_result:
+          break
+        with self._lock:
+          future = self._map[task_result.msgid]
+          del self._map[task_result.msgid]
+          if task_result.success:
+            future.set_result(task_result.value)
+          else:
+            future.set_exception(task_result.value)
+
+      # clear out pending futures and mark ourselves as "stopped" by null-ing
+      # the map
+      with self._lock:
+        for _, v in self._map.items():
+          v.set_exception(concurrent.futures.CancelledError())
+        self._map = None
+
+    def _is_stopped(self):
+      return self._map is None
+
+    def __getattr__(self, name) -> Callable[[Any], concurrent.futures.Future]:
+      result_future = concurrent.futures.Future()
+
+      with self._msgidlock:
+        msgid = self._msgid
+        self._msgid += 1
+
+      def remote_call(*args, **kwargs):
+        with self._lock:
+          if self._is_stopped():
```

Comment:
how about the time gap between the process is shut down but the None info is not received yet? probably just check whether the process is still alive

Reply:
This relies on one signal instead of 2, so less triggers for transitioning the states of the stub. There's not much of an inefficiency happening: if the process was killed and these lines execute first, we just enqueue the message and then the message pump will cancel it. It's all happening at the end of training, so any inefficiency - albeit in this case probably not even measurable - doesn't matter much.